[lldb-dap] Allow providing debug adapter arguments in the extension

matthewbastien opened the pull request:
Added a new setting called `lldb-dap.arguments` and a debug configuration attribute called `debugAdapterArgs` that can be used to set the arguments used to launch the debug adapter. Right now this is mostly useful for debugging purposes to add the `--wait-for-debugger` option to lldb-dap.

Additionally, the extension will now check for a changed lldb-dap executable or arguments when launching a debug session in server mode. I had to add a new `DebugConfigurationProvider` to do this because VSCode will show an unhelpful error modal when the `DebugAdapterDescriptorFactory` returns `undefined`.

In order to facilitate this, I had to add two new properties to the launch configuration that are used by the `DebugAdapterDescriptorFactory` to tell VS Code how to launch the debug adapter:

- `debugAdapterHostname` - the hostname for an existing lldb-dap server
- `debugAdapterPort` - the port for an existing lldb-dap server

I've also removed the check for the `executable` argument in `LLDBDapDescriptorFactory.createDebugAdapterDescriptor()`. This argument is only set by VS Code when the debug adapter executable properties are set in the `package.json`. The LLDB DAP extension does not currently do this (and I don't think it ever will). So, this makes the debug adapter descriptor factory a little easier to read.

The check for whether or not `lldb-dap` exists has been moved into the new `DebugConfigurationProvider` as well. This way the extension won't get in the user's way unless they actually try to start a debugging session. The error will show up as a modal which will also make it more obvious when something goes wrong, rather than popping up as a warning at the bottom right of the screen.

matthewbastien committed:
allow providing debug adapter arguments

JDevlieghere commented on lldb/tools/lldb-dap/package.json:
```diff
@@ -75,6 +75,15 @@
+          "items": {
+            "type": "string"
+          },
+          "description": "The arguments provided to the lldb-dap process."
```
Any reason the description here is different from the ones below? Also, should the description say "The list of _additional_ arguments", as the extension itself might specify others (e.g. with #128957  if you enable server mode)?

matthewbastien replied:
No reason in particular. You're right. I've updated the wording with your suggestions.

JDevlieghere commented on lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:
```diff
@@ -116,40 +131,31 @@ export class LLDBDapDescriptorFactory
     const openSettingsAction = "Open Settings";
+    const message =
+      path === undefined
+        ? "Unable to find the LLDB debug adapter executable."
+        : `Debug adapter path: ${path} is not a valid file`;
```
Either both or neither should have a period at the end (and whichever we pick, it should match what we do elsewhere).

matthewbastien replied:
Good call. This is the only place that we show notifications as far as I can tell. So, I'm leaning towards having the period at the end and changed both to match.

matthewbastien committed:
update wording

JDevlieghere reviewed (approved):
Thanks, this LGTM!

vogelsgesang commented on lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:
```diff
@@ -137,53 +157,59 @@ export class LLDBDapDescriptorFactory
+    args: string[],
+    options: child_process.CommonSpawnOptions,
+  ): Promise<{ host: string; port: number }> {
+    if (this.server) {
+      return this.server;
+    }
```
This does not interact correctly with the new `session.configuration.debugAdapterArgs` settings. We might be reusing a lldb-dap instance which was launched with different arguments, effectively ignoring the `debugAdapterArgs` specified in the launch config.

We already have the same issue with the `dapPath`, where we might end up reusing the wrong lldb-dap binary (see my other comment in https://github.com/llvm/llvm-project/pull/128957#discussion_r1976226941).

But I think this commit is unfortunately further into the direction of incorrectly caching the server processes. Not sure how we want to proceed here...

WDYT, @ashgti and @JDevlieghere ? Is it fine merging this PR as is? Or should we first fix our logic on when to reuse the server and when not?

ashgti replied:
We could prompt the user to ask if they want to stop the running server and start a new server, something like:

```
const stopOption = "Stop running lldb-dap server";
const reuseOption = "Use existing lldb-dap server";
const userInput = await vscode.window.showInformationMessage(
  "An instance of lldb-dap is already running, would you like to stop the running instance or reuse the existing instance?",
   { modal: true },
  stopOption,
  reuseOption,
);

if (!userInput) { 
  return undefined; // This should open the debug configuration
}
if (userInput == stopOption) { ... }
else { ... }
```

vogelsgesang replied:
I assume we would only show this prompt if we detect a mismatch in either the `lldb-dap` binary path or the provided arguments?

JDevlieghere replied:
I think that makes sense 👍

walter-erquinigo replied:
I'm curious. Does this happen in server mode or in all the cases?

Besides that, it would be compare both the dapPath and the arguments when deciding whether to prompt this drop down to the user

matthewbastien replied:
Alright, this took a bit of re-organizing since the `DebugAdapterDescriptorFactory` does not gracefully stop the debug session if it returns `undefined`. Instead, VS Code shows a modal with an error message that suggests that the extension may not have activated properly which is... unhelpful in this case.

I've moved the logic for starting the server and prompting the user for input to a new `DebugConfigurationProvider` which can stop the session gracefully and even open the `launch.json` for editing if something goes wrong.

In order to facilitate this, I had to add two new properties to the launch configuration that are used by the `DebugAdapterDescriptorFactory` to tell VS Code how to launch the debug adapter:

- `debugAdapterHostname` - the hostname for an existing lldb-dap server
- `debugAdapterPort `- the port for an existing lldb-dap server

This has the added bonus of allowing the user to start their own lldb-dap process in server mode if they really wanted to.

vogelsgesang commented on lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:
```diff
@@ -137,53 +157,59 @@ export class LLDBDapDescriptorFactory
    * Shows a message box when the debug adapter's path is not found
    */
-  static async showLLDBDapNotFoundMessage(path?: string) {
+  static async showLLDBDapNotFoundMessage(path?: string | undefined) {
```
Afaict, the `?` already marks `path` as potentially `undefined`? Why add the `| undefined` in addition?

walter-erquinigo replied:
+1

matthewbastien replied:
Done.

vogelsgesang commented on lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:
```diff
@@ -137,53 +157,59 @@ export class LLDBDapDescriptorFactory
         ...env,
       },
     };
-    const dbgArgs = executable?.args ?? [];
```
There might be some downstream users who are actually bundling an `lldb-dap` binary into their re-packaged distribution of this VS-Code extension.

At least the commit message from https://github.com/llvm/llvm-project/commit/b5d4332286154838557a8ab5c76b794e85d946b3 indicates, that there were some thoughts that downstream consumers might inject the lldb-dap path when using this extension as a library.

@JDevlieghere @walter-erquinigo are you aware if anyone is re-packaging this VS-Code extension together with an lldb-dap binary and might be relying on setting the lldb-dap path through the `package.json`?

JDevlieghere replied:
I'm not but I'd be curious to find out too.

walter-erquinigo replied:
Long story short, I don't think anyone ever did that.

I intended to do that when I was working on the mojo extension, but I never really do that because I was waiting for the upstream code to have a feature the mojo extension didn't have in order to do the switch, but that didn't happen last year. This year I'm not working on that company anymore, so I don't think anyone will do that in the foreseable future. I know that Meta has their own debugger extension, which is very big, and at least they shouldn't be repackaging it. I don't know about other companies

walter-erquinigo replied:
But anyway, I'm fine with this change if we want less ways to set the args, which is good. However, you should throw an exception if `executable` is not `undefined`, which would be an unintended flow that might not work the way users expect.

vogelsgesang replied:
ok, if you are fine with it, I am also fine with it

matthewbastien replied:
Alright, this took a bit of re-organizing since the `DebugAdapterDescriptorFactory` does not gracefully stop the debug session if it returns `undefined`. Instead, VS Code shows a modal with an error message that suggests that the extension may not have activated properly which is... unhelpful in this case.

I've moved the logic for starting the server and prompting the user for input to a new `DebugConfigurationProvider` which can stop the session gracefully and even open the `launch.json` for editing if something goes wrong.

In order to facilitate this, I had to add two new properties to the launch configuration that are used by the `DebugAdapterDescriptorFactory` to tell VS Code how to launch the debug adapter:

- `debugAdapterHostname` - the hostname for an existing lldb-dap server
- `debugAdapterPort `- the port for an existing lldb-dap server

This has the added bonus of allowing the user to start their own lldb-dap process in server mode if they really wanted to.

The `DebugAdapterDescriptorFactory` will also throw an error if the executable parameter is set.

walter-erquinigo commented on lldb/tools/lldb-dap/package.json:
```diff
@@ -162,6 +171,13 @@
+                "items": {
+                  "type": "string"
+                },
+                "markdownDescription": "The list of additional arguments used to launch the debug adapter executable."
```
Mention that this will override the VSCode-level settings for launching the adapter

matthewbastien replied:
Done.

walter-erquinigo commented on lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:
```diff
@@ -93,13 +93,33 @@ async function getDAPExecutable(
+function getDAPArguments(session: vscode.DebugSession): string[] {
+  // Check the debug configuration for arguments first
+  const debugConfigArgs = session.configuration.debugAdapterArgs;
+  if (
+    Array.isArray(debugConfigArgs) &&
+    debugConfigArgs.findIndex((entry) => typeof entry !== "string") === -1
+  ) {
+    return debugConfigArgs;
+  }
```
Don't be silent about errors. If the `debugConfigArgs` variable is not `undefined` and has unexpected data, then just throw an exception. Otherwise the user won't know that something wrong happened and they will waste time.

matthewbastien replied:
Added an error message if `debugAdapterArgs` is not in the expected format.

matthewbastien committed:
prompt the user to restart the server if the executable or arguments change

matthewbastien committed:
add more checks and error messages

matthewbastien committed:
mention that debug configuration properties override VS Code settings

walter-erquinigo commented on lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:
```diff
@@ -66,19 +70,17 @@ async function findDAPExecutable(): Promise<string | undefined> {
 
 async function getDAPExecutable(
-  session: vscode.DebugSession,
+  folder: vscode.WorkspaceFolder | undefined,
```
call this workspaceFolder because this is not any folder

matthewbastien replied:
Done.

matthewbastien committed:
tell the user if lldb-dap doesn't exist when a debug session is started

matthewbastien committed:
change `folder` arguments to `workspaceFolder`

matthewbastien committed:
show more detailed error messages depending on what went wrong

matthewbastien commented:
I refactored the UI logic so that we don't have to pass `userInteractive` everywhere and can share common buttons with different parts of the app. I also added where the lldb-dap path comes from in the error message.

Probably worth another review before merging.

CC @JDevlieghere @walter-erquinigo

JDevlieghere commented on lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:
```diff
@@ -90,117 +99,119 @@ async function getDAPExecutable(
+    }
+    return debugConfigArgs;
+  }
+  // Fall back on the workspace configuration
```
```suggestion
  // Fall back on the workspace configuration.
```

JDevlieghere commented on lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:
```diff
@@ -90,117 +99,119 @@ async function getDAPExecutable(
+  workspaceFolder: vscode.WorkspaceFolder | undefined,
+  configuration: vscode.DebugConfiguration,
+): Promise<string[]> {
+  // Check the debug configuration for arguments first
```
```suggestion
  // Check the debug configuration for arguments first.
```

JDevlieghere reviewed (approved):
I'd like @vogelsgesang and @ashgti to have a look before we merge this, but the new experience seems really nice.

I left a few nits about comments missing a period. Event though this is TypeScript, we should follow the [LLVM Coding Standards](https://llvm.org/docs/CodingStandards.html) unless there's a good reason to diverge.

matthewbastien committed:
fix some review comments

ashgti commented on lldb/tools/lldb-dap/src-ts/lldb-dap-server.ts:
```diff
@@ -0,0 +1,130 @@
+import * as child_process from "node:child_process";
+import * as vscode from "vscode";
+
+function areArraysEqual<T>(lhs: T[], rhs: T[]): boolean {
+  if (lhs.length !== rhs.length) {
+    return false;
+  }
+  for (let i = 0; i < lhs.length; i++) {
+    if (lhs[i] !== rhs[i]) {
+      return false;
+    }
+  }
+  return true;
+}
```
nit: Nodejs does include a utility for this https://nodejs.org/docs/latest/api/util.html#utilisdeepstrictequalval1-val2

ashgti commented on lldb/tools/lldb-dap/src-ts/lldb-dap-server.ts:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    const userInput = await vscode.window.showInformationMessage(
+      "A server mode instance of lldb-dap is already running, but the arguments are different from what is requested in your debug configuration or settings. Would you like to restart the server?",
```
Should we have a shorter message title and use the detail in the options? e.g. 

```ts
await vscode.window.showInformationMessage(
 'lldb-dap arguments have changed, would you like to restart the server?'
 { 
    modal: true, 
    detail: `An existing lldb-dap server (${this.serverProcess.pid}) is running with different arguments.

The previous lldb-dap server was started with:

${this.serverProcess.spawnargs.joined(' ')}

Restarting the server will interrupt any existing debug sessions and start a new server.`
 },
 "Restart",
  "Use Existing",
 )`

matthewbastien replied:
I like this way better because the modal reads nicer this way as well. Done.

ashgti reviewed (approved):
Overall looks great, a few questions / follow items

matthewbastien committed:
use resolveDebugConfigurationWithSubstitutedVariables

matthewbastien committed:
move isExecutable() error logic into getDAPExecutable()

matthewbastien committed:
use util.isDeepStrictEqual() instead of custom equality logic

matthewbastien committed:
update wording in server restart modal

matthewbastien commented on lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:
```diff
@@ -90,117 +99,119 @@ async function getDAPExecutable(
+      buttons = [new ConfigureButton()];
+    }
+    throw new ErrorWithNotification(message, ...buttons);
+  }
```
Good point. I've moved this logic into `getDAPExecutable()`.

matthewbastien commented on lldb/tools/lldb-dap/src-ts/debug-configuration-provider.ts:
```diff
@@ -0,0 +1,103 @@
+    folder: vscode.WorkspaceFolder | undefined,
+    debugConfiguration: vscode.DebugConfiguration,
+    _token?: vscode.CancellationToken,
+  ): Promise<vscode.DebugConfiguration | null | undefined> {
```
Done.

matthewbastien commented on lldb/tools/lldb-dap/src-ts/debug-configuration-provider.ts:
```diff
@@ -0,0 +1,103 @@
+
+  async resolveDebugConfiguration(
+    folder: vscode.WorkspaceFolder | undefined,
+    debugConfiguration: vscode.DebugConfiguration,
```
I really like this idea!

Playing around with your code locally, I found that we would also need some more complex validators for things like making sure `debugAdapterPort` and `debugAdapterExecutable` aren't set together. Would also be nice to have some of the validators coerce strings to numbers for things like the port information which accept both.

I think this definitely warrants a separate PR that consolidates the validation logic.
